classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,218 @@
+        if (is_null($mergerequest->keepuserid)) {
+            $this->keepuserid = $this->find_user_id_or_fail($mergerequest->keepuserfield,
+                                                                $mergerequest->keepuservalue,
+                                                                $mergerequest->retries
+                                                            );
+        }
+    }
+    /**
+     * Function for updating number of retries.
+     */
+    private function update_retries(object $mergerequest): void {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $mergerequest->id,
+                'retries' => $mergerequest->retries + 1,
+                'timecompleted' >= time(),
+            ]
+        );
+    }
+    /**
+     * Function for updating removeuserid.
+     */
+    private function update_removeuserid_in_table(int $idrecord,
+                                                int $removeuserid): void {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'removeuserid' => $removeuserid
+            ]
+        );
+    }
+    /**
+     * Function for updating keepuserid.
+     */
+    private function update_keepuserid_in_table(int $idrecord,
+                                                int $keepuserid): void {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'keepuserid' => $keepuserid
+            ]
+        );
+    }
+    /**
+     * Function to find user id or fail.
+     */
+    protected function find_user_id_or_fail(string $userfield,
+                                            string $uservalue,
+                                            int $retries
+                                            ): int {
+        global $DB;
+        $users = $DB->get_records(merge_request::TABLE_USERS, [$userfield => $uservalue]);
+        $errorfound = '';
+        if (count($users) == 0) {
+            $errorfound = get_string(get_string('cannotfinduser', 'tool_mergeusers',
+                                                (object)['userfield' => $userfield,
+                                                        'uservalue' => $uservalue,
+                                                        'retries' => $retries,
+                                                        ]));
+        } else if (count($users) > 1) {
+            $errorfound = get_string(get_string('toomanyusers', 'tool_mergeusers',
+            (object)['userfield' => $userfield,
+                    'uservalue' => $uservalue,
+                    'retries' => $retries,
+                    ]));
+        }
+        if (!empty($errorfound)) {
+            $mergerequest->appendlog([$errorfound], time());
```

Comment:
there is no `$mergerequest` initialized here in this context. Review the code, please.

Reply:
still pending.